src/Couchbase.Lite.Shared/API/Database/Database.cs

```diff
@@ -816,10 +816,55 @@ public void RemoveChangeListener(ListenerToken token)
         public bool Save([NotNull]MutableDocument document, ConcurrencyControl concurrencyControl)
         {
             var doc = CBDebug.MustNotBeNull(WriteLog.To.Database, Tag, nameof(document), document);
-            return Save(doc, concurrencyControl, false);
+            return Save(doc, null, concurrencyControl, false);
         }
 
-        #if CBL_LINQ
+        /// <summary>
+        /// Saves a document to the database. When write operations are executed concurrently, 
+        /// and if conflicts occur, conflict handler will be called. Use the handler to directly
+        /// edit the document.Returning true, will save the document. Returning false, will cancel
+        /// the save operation.
+        /// </summary>
+        /// <param name="document">The document to save</param>
+        /// <param name="conflictHandler">The conflict handler block which can be used to resolve it.</param> 
+        /// <returns><c>true</c> if the save succeeded, <c>false</c> if there was a conflict</returns>
+        public bool Save([NotNull]MutableDocument document, Func<MutableDocument, Document, bool> conflictHandler)
+        {
+            var doc = CBDebug.MustNotBeNull(WriteLog.To.Database, Tag, nameof(document), document);
+            Debug.Assert(conflictHandler != null);
+            Document currDoc = null;
+            while (true) {
+                var saved = Save(doc, currDoc, ConcurrencyControl.FailOnConflict, false);
+                if (!saved) { // has conflict, save failed
+                    var oldDoc = GetDocument(doc.Id);
+                    C4Error err;
+                    C4Document* curDoc = Native.c4doc_get(_c4db, doc.Id, true, &err);
+                    var isDeleted = curDoc->flags.HasFlag(C4DocumentFlags.DocDeleted);
+                    if (isDeleted) {
+                        oldDoc.ReplaceC4Doc(new C4DocumentWrapper(curDoc));
+                        Native.c4doc_free(curDoc);
+                        curDoc = null;
+                    }
+                    if (oldDoc == null) {
+                        return false;
+                    }
+                    currDoc = isDeleted ? null : oldDoc;
```

Comment:
I can't understand what is going on with any of the lines between `if(!saved)` and here...

What is their purpose?

Also `curDoc` will be leaked if `isDeleted` is `false`

Reply:
I know it's confusing but curDoc and currDoc are different. I freed curDoc (a C4Document) on line 845 and 845.

Reply:
are the better names the curDoc and currDoc? I can easily see someone misreading one of the "r"'s and getting confused.